tests/network/vmi_istio.go

```diff
@@ -180,42 +216,70 @@ var _ = SIGDescribe("[Serial] Istio", func() {
+					&expect.BExp{R: returnCode},
 					&expect.BSnd{S: "echo $?\n"},
 					&expect.BExp{R: console.RetValue("0")},
 				}, 60)
-				Expect(err).ToNot(HaveOccurred())
+			}
+
+			Context("VMI with explicit ports", func() {
+				BeforeEach(func() {
+					vmiPorts = explicitPorts
+				})
+				It("Should be able to reach http server outside of mesh", func() {
+					Expect(
+						checkHTTPServiceReturnCode(serverVMIAddress, testPort, generateExpectedHTTPReturnCodeRegex("200")),
+					).ToNot(HaveOccurred())
+				})
+			})
+			Context("VMI with no explicit ports", func() {
+				BeforeEach(func() {
+					vmiPorts = []v1.Port{}
+				})
+				It("Should be able to reach http server outside of mesh", func() {
+					Expect(
+						checkHTTPServiceReturnCode(serverVMIAddress, testPort, generateExpectedHTTPReturnCodeRegex("200")),
+					).ToNot(HaveOccurred())
+				})
 			})
+
 			Context("With Sidecar allowing only registered external services", func() {
+				// Istio Envoy will intercept the request because of the OutboundTrafficPolicy set to REGISTRY_ONLY.
+				// Envoy responds with 502 Bad Gateway return code.
+				// After Sidecar with OutboundTrafficPolicy is created, it may take a while for the Envoy proxy
+				// to sync with the change, first request may still get through, hence the Eventually used for assertions.
+
 				BeforeEach(func() {
 					sidecarRes := schema.GroupVersionResource{Group: "networking.istio.io", Version: "v1beta1", Resource: "sidecars"}
 					registryOnlySidecar := generateRegistryOnlySidecar()
 					_, err = virtClient.DynamicClient().Resource(sidecarRes).Namespace(tests.NamespaceTestDefault).Create(context.TODO(), registryOnlySidecar, metav1.CreateOptions{})
 					Expect(err).ShouldNot(HaveOccurred())
 				})
 
-				// Istio Envoy will intercept the request because of the OutboundTrafficPolicy set to REGISTRY_ONLY.
-				// Envoy responds with 502 Bad Gateway return code.
-				// After Sidecar with OutboundTrafficPolicy is created, it may take a while for the Envoy proxy
-				// to sync with the change, first request may still get through, hence the Eventually.
-				It("Should not be able to reach http server outside of mesh", func() {
-					Eventually(func() error {
-						return console.SafeExpectBatch(vmi, []expect.Batcher{
-							&expect.BSnd{S: "\n"},
-							&expect.BExp{R: console.PromptExpression},
-							&expect.BSnd{S: curlCommand(serverVMIAddress, testPort)},
-							&expect.BExp{R: generateExpectedHTTPReturnCodeRegex("5..")},
-							&expect.BSnd{S: "echo $?\n"},
-							&expect.BExp{R: console.RetValue("0")},
-						}, 60)
-					}, 90*time.Second, 10*time.Second)
+				Context("VMI with explicit ports", func() {
+					BeforeEach(func() {
+						vmiPorts = explicitPorts
+					})
+					It("Should not be able to reach http service outside of mesh", func() {
+						Eventually(func() error {
+							return checkHTTPServiceReturnCode(serverVMIAddress, testPort, generateExpectedHTTPReturnCodeRegex("5.."))
```

Comment:
Why is this one wrapped in an `Eventually` block ?

Reply:
The reason for the `Eventually` is that it takes a little bit of time for the proxy to sync with the daemon about the newly created Sidecar, that blocks external services.
Which means the first request may still get through.
The timeouts are overkill though

Reply:
> The reason for the `Eventually` is that it takes a little bit of time for the proxy to sync with the daemon about the newly created Sidecar, that blocks external services.
> Which means the first request may still get through.
> The timeouts are overkill though

Is that readiness state capture somewhere ? I.e. would it be possible to wait until a certain even and then just try once, rather than have the eventually block ?

Not saying it must be that way; it's good enough as it currently is. But we would be better off if such a thing was possible.

Reply:
We could check what they are doing there, and encapsulate the required subset of the implementation so we know the proxy is in synch with istiod, and only then check the traffic behaves as we intend.

This is not (at all) required in this PR, but I think it would be a good overall improvement for the future.